File: solutions/javascript/GuessNumberHigherorLowerII/GuessNumberHigherorLowerII.js
/**
 * @param {number} n
 * @return {number}
 */
var getMoneyAmount = function(n) {
    let table = [];
    for (let i = 0; i <= n; i++) {
        table[i] = Array(n + 1).fill(0);
    }
    return xxx(table, 1, n);
};
var xxx = function(t, s, e) {
    if (s >= e) {
        return 0;
    }
    if (t[s][e] !== 0) {
        return t[s][e];
    }
    let res = Infinity;
    for (let x = s; x <= e; x++) {
        let tmp = x + Math.max(xxx(t, s, x - 1), xxx(t, x + 1, e));
        res = Math.min(res, tmp);
    }
    t[s][e] = res;
    return res;
};

File: solutions/java/GuessNumberHigherorLowerII/GuessNumberHigherorLowerII.java
class Solution {
    public int getMoneyAmount(int n) {
        if (n <= 1) {
            return 0;
        }
        int[][] dp = new int[n + 1][n + 1];
        for (int r = 2; r <= n; ++r) {
            for (int l = r - 1; l > 0; --l) {
                if (l + 1 == r) dp[l][r] = l;
                else if (l + 2 == r) dp[l][r] = l + 1;
                else {
                    dp[l][r] = Integer.MAX_VALUE;
                    for (int k = l + 2; k < r; ++k) {
                        dp[l][r] = Math.min(dp[l][r], k + Math.max(dp[l][k - 1], dp[k + 1][r]));
                    }
                }
            }
        }
        return dp[1][n];
    }
}